src/lib/FreeC/LiftedIR/Converter/Expr.hs

```diff
@@ -0,0 +1,190 @@
+  ( convertExpr
+  )
+where
+
+import           Control.Monad                  ( join )
+import           Data.Maybe                     ( fromMaybe )
+
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.IR.SrcSpan               ( SrcSpan(NoSrcSpan) )
+import qualified FreeC.LiftedIR.Syntax         as LIR
+import qualified FreeC.LiftedIR.Converter.Type as LIR
+import           FreeC.Environment
+import           FreeC.Environment.LookupOrFail ( lookupAgdaFreshIdentOrFail
+                                                , lookupAgdaValIdentOrFail
+                                                )
+import           FreeC.Environment.Renamer      ( renameAndDefineAgdaVar )
+import           FreeC.Environment.Fresh        ( freshIRQName )
+import           FreeC.Monad.Converter
+import           FreeC.Monad.Reporter           ( reportFatal
+                                                , Message(Message)
+                                                , Severity(Internal)
+                                                )
+
+-- | Converts an expression from IR to lifted IR and lifts it during the
+--   translation.
+convertExpr :: IR.Expr -> Converter LIR.Expr
+convertExpr expr = convertExpr' expr [] []
+
+-- | Same as @convertExpr@ but accumulates arguments.
+--
+--   This function always produces a term of type @Free S P τ*@.
+--   The accumulation of arguments is needed to reason about fully applied
+--   functions. Top level functions and constructors are lifted argument wise.
+--   Translating them with out considering all arguments would violate the
+--   invariant.
+--
+--   > e : τ ↦ e' : τ'
+convertExpr' :: IR.Expr -> [IR.Type] -> [IR.Expr] -> Converter LIR.Expr
+
+-- Pass argument from applications to converter for callee, allowing us to
+-- convert functions and constructors with full access to their parameters.
+--
+-- >                $
+-- > convertExpr'  / \   [] args = convertExpr' e₁ [] (e₂ : args)
+-- >              e₁  e₂
+convertExpr' (IR.App _ e1 e2 _) [] args = convertExpr' e1 [] (e2 : args)
+
+-- Pass type argument from visible type application to converter for callee.
+--
+-- >                @
+-- > convertExpr'  / \   tArgs args = convertExpr' e (τ : tArgs) args
+-- >              e   τ
+convertExpr' (IR.TypeAppExpr _ e t _) typeArgs args =
+  convertExpr' e (t : typeArgs) args
+
+-- Constructors.
+convertExpr' (IR.Con srcSpan name _) _ args = do
+  args' <- mapM convertExpr args
+  let con = LIR.SmartCon srcSpan name
+  return $ LIR.App srcSpan con [] [] args'
+
+-- Cases for (possible applied) variables (i.e. variables and functions).
+convertExpr' (IR.Var srcSpan name _) _ args = do
+  args'    <- mapM convertExpr args
+  varName  <- LIR.Var srcSpan name <$> lookupAgdaValIdentOrFail srcSpan name
+  function <- inEnv $ isFunction name
+  if function -- If this is a top level functions it's lifted argument wise.
+    then return $ LIR.App srcSpan varName [] [] args'
```

Comment:
I'm not entirely sure, but shouldn't the bind generation for strict arguments already happen during the IR to LIR transformation? Otherwise new binds have to be generated by the Coq backend.

The same problem occurs during `convertAlt`. Here the strictness information in the `IR.VarPat`s gets lost, which makes the LIR to Coq transformation somewhat difficult.